markwon/src/main/java/ru/noties/markwon/renderer/SpannableBuilder.java

```diff
@@ -0,0 +1,63 @@
+package ru.noties.markwon.renderer;
+
+import android.support.annotation.Nullable;
+import android.text.SpannableStringBuilder;
+
+/**
+ * Copied with modification and renaming from @see <a href="https://github.com/Uncodin/bypass/blob/master/platform/android/
+ * library/src/in/uncod/android/bypass/ReverseSpannableStringBuilder.java">Uncodin/bypass</a>
+ */
+public class SpannableBuilder extends SpannableStringBuilder {
+
+    public SpannableBuilder() {
+        super();
+    }
+
+    SpannableBuilder(CharSequence text, int start, int end) {
+        super(text, start, end);
+    }
+
+    @Override
+    public <T> T[] getSpans(int queryStart, int queryEnd, Class<T> kind) {
+        T[] ret = super.getSpans(queryStart, queryEnd, kind);
+        reverse(ret);
+        return ret;
+    }
+
+    @Override
+    public void setSpan(@Nullable Object what, int start, int end, int flags) {
+        if (what != null && isRangeValid(start, end)) {
+            if (what.getClass().isArray()) {
+                for (final Object span : (Object[]) what) {
+                    setSpan(span, start, end, flags);
+                }
+            } else {
+                super.setSpan(what, start, end, flags);
+            }
+        }
+    }
+
+    private boolean isRangeValid(int start, int end) {
+        final int len;
```

Comment:
Okay, I see what you are doing, but we do not need to lazily introduce `length` variable. It's not worth it. 

I also think that there are some logic flows...
* `end` must be **greater** that `start`: `end > start`
* `start` must be **less** `len`: `start < len`
* we do not need last `end >= 0` check (which should be btw `end >0`), because we have already checked if end is greater than start.

So, something like this would be better:
```java
final int length = length();
return end > start
    && start >= 0
    && start < length
    && end <= length;
```

Reply:
And yeah... let's make `len` a grown-up and give her a proper name: `length`

Reply:
I actually cheated and used the logic from `SpannableStringBuilder`'s range check:

```java
private void checkRange(final String operation, int start, int end) {
    if (end < start) {
        throw new IndexOutOfBoundsException(operation + " " +
                region(start, end) + " has end before start");
    }

    int len = length();

    if (start > len || end > len) {
        throw new IndexOutOfBoundsException(operation + " " +
                region(start, end) + " ends beyond length " + len);
    }

    if (start < 0 || end < 0) {
        throw new IndexOutOfBoundsException(operation + " " +
                region(start, end) + " starts before 0");
    }
}
```

Based on the above I originally wrote my range check it as:
```java
if (end < start || (start > (len = length()) || end > len) || (start < 0 || end < 0)) {
    return false;
}
return true;
```

Reply:
Actually:

```java
final int length = length();
return end > start
    && start >= 0
    && end <= length;
```

will do, as `start` will be less than length anyway